File: DynamicMapping/dynamic_mapping/geometry_maps/primitives/circle.py
import numpy as np

# from scene.scene import Scene

class Circle:

    def __init__(self, xc, yc, R):
        """
        Instantiates a circle to represent a cluster of points
        
            :param xc: x-position of the center of the circle in meters
            :param yc: y-position of the center of the circle in meters
            :param R: Radius of the circle in meters
        """
        
        # Center and radius of the circle:
        self.xc = xc
        self.yc = yc
        self.R = R

        # Derivative of xc and yc (used to estimate the speed of the circle):
        self.d_xc = 0
        self.d_yc = 0

        # If the circle is consistent with the current observations:
        self.is_valid = True

        # Color of the circle (for drawing):
        self.circle_color = (255, 0, 0)

    def draw(self, scene : 'Scene', draw_speed_estimate : bool = False):
        """ Draws the circle in the scene """

        scene.draw_circle(self.xc, self.yc, self.R, self.circle_color)

        if draw_speed_estimate:
            angle = np.arctan2(self.d_yc, self.d_xc)
            length = np.sqrt(self.d_xc ** 2 + self.d_yc ** 2)
            scene.draw_arrow(self.xc, self.yc, self.R, length, angle, (255, 0, 0))

    def update_circle_using_matching(self, other : 'Circle'):
        """
        Supposing that this circle belongs to the current model of the environment, use the
        given circle (supposed to be an observed circle, matched with this one) to update the
        center and radius of this circle
        """

        new_xc = (self.xc + other.xc) / 2
        new_yc = (self.yc + other.yc) / 2
        self.R = (self.R + other.R) / 2

        # Use the new_xc and new_yc to update the circle speed estimate, using a simple
        # exponential low pass filter (TODO: Use Kalman Filter instead):
        m = 0.95
        self.d_xc = m * self.d_xc + (1 - m) * (new_xc - xc)
        self.d_yc = m * self.d_yc + (1 - m) * (new_yc - yc)

        # Update xc and yc:
        xc = new_xc
        yc = new_yc

    def distance_from(self, other : 'Circle'):
        """
        Computes the Euclidean distance between the center of this circle and
        the center of the given circle
        """

        dX = other.xc - self.xc
        dY = other.yc - self.yc

        return np.Sqrt(dX * dX + dY * dY)

File: IDP - Dynamic Mapping/Assets/Scripts/Mapping/GeometryClustering/Primitives/Circle.cs
using System.Collections.Generic;
using System.Net;
using UnityEngine;

/// <summary>
/// Class used to represent circle primitives for the mapping using geometric primitives
/// </summary>
public class Circle {
    
    // Color of the circle (for drawing):
    public Color circleColor = Color.red;

    // Center and radius of the circle:
    private Vector2 _center;
    public Vector2 center { get => _center; }
    private float R;

    // Estimated derivative of center.x and center.y with respect to time:
    private Vector2 speed;

    // If the circle is consistent with the current observations:
    public bool isValid = true;

    /// <summary>
    /// Instantiates a circle with the given center and radius
    /// </summary>
    /// <param name="center">World position of the circle</param>
    /// <param name="R">Radius of the circle in meters</param>
    public Circle(Vector2 center, float R) {
        this._center = center;
        this.R = R;

        this.speed = Vector2.zero;
    }

    /// <summary>
    /// Draws the circle in the scene using Unity gizmos
    /// </summary>
    /// <param name="height">Position of the circle along the Z-axis</param>
    public void DrawGizmos(float height, bool drawSpeedEstimate) {
        Gizmos.color = circleColor;
        Gizmos.DrawSphere(Utils.To3D(_center, height), R);

        // Draw an arrow representing the estimated speed of the circle:
        if (drawSpeedEstimate) {
            Vector3 p1 = Utils.To3D(_center, height);
            Vector3 p2 = Utils.To3D(_center + 10 * speed, height);

            Gizmos.color = Color.red;
            Gizmos.DrawLine(p1, p2);
        }
    }

    /// <summary>
    /// From the previous circle estimate and the elapsed time since the last update,
    /// predict where the circle should be now
    /// </summary>
    /// <param name="elapsedTime">Elapsed time in seconds since the last update</param>
    /// <param name="friction">Friction applied to the speed of the circle (between 0 and 1)</param>
    public void PredictState(float elapsedTime, float friction) {
        _center += elapsedTime * speed;
        speed *= 1 - friction;
    }

    /// <summary>
    /// Supposing that this circle belongs to the current model of the environment, use the
    /// given circle (that is supposed to be matched with this one) to update the center and
    /// radius of this circle
    /// </summary>
    /// <param name="other">Observed circle matched with this one</param>
    public void UpdateCircleUsingMatch(Circle other) {
        Vector2 newCenter = (_center + other._center) / 2;
        R = (R + other.R) / 2;

        // Use the newCenter to update the circle speed estimate, using a simple
        // exponential low pass filter (TODO: Implement Kalman Filter instead):
        const float m = 0.9f;
        speed = m * speed + (1 - m) * (newCenter - _center);

        // Update the center of the circle:
        _center = newCenter;
    }

    /// <summary>
    /// Computes the Euclidean distance between the centers of this circle and the
    /// center of the given circle
    /// </summary>
    public float DistanceFrom(Circle other) {
        return (other._center - _center).magnitude;
    }

    /// <summary>
    /// Returns if the circle is far enough from the given lines
    /// </summary>
    /// <param name="lines">List of lines</param>
    /// <param name="minDistanceToLines">Minimum allowed orthogonal distance 
    /// between a line and the center of the circle</param>
    /// <returns></returns>
    public bool IsFarFromLines(List<DynamicLine> lines, float minDistanceToLines) {
        if (lines == null)
            return true;

        foreach (DynamicLine line in lines)
            if (line.DistanceOf(_center) < minDistanceToLines)
                return false;

        return true;
    }
}
